File: Java/706. Design HashMap.java
class MyHashMap {
    private int range;
    private List<Bucket> buckets;

    public MyHashMap() {
        this.range = 2069;
        this.buckets = new ArrayList<Bucket>();
        for(int i=0; i<this.range; i++){
            this.buckets.add(new Bucket());
        }
    }
    
    public void put(int key, int value) {
        int index = key % this.range;
        this.buckets.get(index).insert(key, value);
    }
    
    public int get(int key) {
        int index = key % this.range;
        return this.buckets.get(index).search(key);
    }
    
    public void remove(int key) {
        int index = key % this.range;
        this.buckets.get(index).delete(key);
    }
}

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap obj = new MyHashMap();
 * obj.put(key,value);
 * int param_2 = obj.get(key);
 * obj.remove(key);
 */

class Bucket {
    private List<Pair<Integer, Integer>> bucket;
    
    public Bucket() {
        this.bucket = new LinkedList<Pair<Integer, Integer>>();
    }
    
    // search a key value pair
    public Integer search(Integer key){
        for(Pair<Integer, Integer> pair: this.bucket){
            if(pair.first.equals(key)){
                return pair.second;
            }
        }
        return -1;
    }
    
    // insert a key value pair or update the value
    public void insert(Integer key, Integer val){
        boolean found = false;
        for(Pair<Integer, Integer> pair: this.bucket){
            if(pair.first.equals(key)){
                pair.second = val;
                found = true;
                break;
            }
        }
        if(!found){
            this.bucket.add(new Pair<Integer, Integer>(key, val));
        }
    }
    
    // delete a key value pair
    public void delete(Integer key){
        for(Pair<Integer, Integer> pair: this.bucket){
            if(pair.first.equals(key)){
                this.bucket.remove(pair);
                break;
            }
        }
    }
}

class Pair<U, V> {
    public U first;
    public V second;
    
    public Pair(U first, V second) {
        this.first = first;
        this.second = second;
    }
}

File: Python3/706. Design HashMap.py
class MyHashMap:

    def __init__(self):
        self.range = 2069
        self.buckets = [Bucket() for _ in range(self.range)]

    def put(self, key: int, value: int) -> None:
        index = key % self.range
        self.buckets[index].insert(key, value)

    def get(self, key: int) -> int:
        index = key % self.range
        return self.buckets[index].search(key)

    def remove(self, key: int) -> None:
        index = key % self.range
        self.buckets[index].delete(key)


# Your MyHashMap object will be instantiated and called as such:
# obj = MyHashMap()
# obj.put(key,value)
# param_2 = obj.get(key)
# obj.remove(key)

class Bucket:
    def __init__(self):
        # basic data structure is a tuple: (key, value)
        self.bucket = []
    
    # insert a (key, value) pair if it doesn't exist
    # otherwise, update the value
    def insert(self, key, val):
        found = False
        for i, pair in enumerate(self.bucket):
            if key == pair[0]:
                found = True
                # update the value
                self.bucket[i] = (key, val)
                break
                
        # if not found, append the (key, val) pair
        if not found:
            self.bucket.append((key, val))
            
    def search(self, key):
        for k, v in self.bucket:
            if k == key:
                return v
        return -1
    
    def delete(self, key):
        for i, pair in enumerate(self.bucket):
            if key == pair[0]:
                del self.bucket[i]
        